README.md
# FeatSeekR - Package for unsupervised feature selection

A fundamental step in many analyses of high-dimensional data is dimension 
reduction. Feature selection is one approach to dimension reduction whose 
strengths include interpretability, conceptual simplicity, transferability 
and modularity.
Here, we introduce the `FeatSeekR` package, which selects features based on 
the consistency of their signal across replicates and their non-redundancy.
It takes a 2 dimensional array (features x samples) of replicated measurements
and returns a `SummarizedExperiment` object storing the selected and features ranked by 
reproducibility. This work was motivated by [[1]](#1) who devised a special case of
our current method to use it on microscopy data, but did not implement it as R package.


# Installation


```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("FeatSeekR")
```


# How to run

See vignette for more detailed examples.

```{r}
set.seed(111)
# simulate data with 500 samples, 3 replicates and 5 latent factors 
# generating 50 features
samples <- 500
latent_factors <- 5
replicates <- 3

sim <- FeatSeekR::simData(samples=samples,latent_factors =latent_factors,
                replicates = replicates)
data <- sim[[1]]
reps <- sim[[2]]

# select the top 5 features
res <- FeatSeek(data, replicates =reps, max_features=5)

# plot a heatmap of the top 5 selected features 
FeatSeekR::plotSelectedFeatures(data, res)
```


## References
<a id="1">[1]</a> 
Fischer, B., Sandmann, T., Horn, T., Billmann, M., Chaudhary, V., Huber, W. and Boutros, M., 2015. A map of directional genetic interactions in a metazoan cell. Elife, 4, p.e05464.
```


NEWS
Changes in version 0.99.1 
+ Submitted to Bioconductor


tests/testthat.R
library(testthat)
library(FeatSeekR)

test_check("FeatSeekR")


tests/testthat/test-plots.R
test_that("plotSelectedFeatures returns plot",{
    data <-  array(rnorm(50*30), dim=c(30,50),
                dimnames=list(paste("feature", seq_len(30)), NULL))
    reps <- rep(c(1,2), each=25)

    selected <- paste("feature", seq_len(5))
    res <- data.frame(selected = selected)

    se_res <- SummarizedExperiment::SummarizedExperiment(
        assays=list(selected=data[selected,]),
        rowData=res)

    p <- plotSelectedFeatures(se_res)
    
    expect_error(plotSelectedFeatures(data))
    expect_error(plotSelectedFeatures(se_res, n_features=c("feature 1", "feature 2")))
    expect_error(plotSelectedFeatures(se_res, assay="sel_assay"))
    
    expect_true(all(p$tree_row$labels == selected))
    expect_true(is(p,"pheatmap"))
})

test_that("plotVarianceExplained returns plot",{
    data <-  array(rnorm(50*30), dim=c(30,50),
                   dimnames=list(paste("feature", seq_len(30)), NULL))
    reps <- rep(c(1,2), each=25)
    
    selected <- paste("feature", seq_len(5))
    res <- data.frame(selected = selected)
    se <- SummarizedExperiment::SummarizedExperiment(
      assays=list(selected=data[selected,]),
      rowData=res)
    
    se_res <- FeatSeek(data, reps, max_features=5)
   
    expect_true(is.null(plotVarianceExplained(se_res)))
    expect_error(plotVarianceExplained(data))
    expect_error(plotVarianceExplained(se), "'res' does not contain attribute explained_variance. Please run FeatSeek first!")
    
})
  

tests/testthat/test-utils.R


test_that("calcFstat returns correct output",{
    data <-  array(rnorm(100*3), dim=c(100,3), dimnames=list(NULL, c("first", "second", "third")))
    reps <- rep(c(1,2), each=50)
    Fstat <- calcFstat(data, reps, scale=TRUE)
    
    expect_error(calcFstat(data, c(1,2), scale=TRUE))
    expect_error(calcFstat(data, reps, scale="yes"))
    
    expect_true(is(Fstat, "numeric"))
    expect_identical(length(Fstat), dim(data)[[2]])
})


R/fit_lm.R
#' @title fit_lm
#'
#' @description Fit linear model for each feature as a function of
#' the previously selected features S.
#' The dimensions spanned by the selected features are
#' projected out of the data by setting each feature
#' to its residuals from the linear model fit.
#'
#' @param data \code{data} (2 dimensional array samples x features)
#' @param S set of selected features
#' @param k current iteration
#'
#' @return \code{data} with previously selected features projected out
#' 
#' @importFrom stats lm na.exclude
#' @importFrom methods is
#' @keywords internal
fit_lm <- function(data, S, k){
    stopifnot(
        is.numeric(data),
        is.numeric(S),
        k-1 <= ncol(S),
        length(k) == 1
    )
    # get current features in data
    features <- dimnames(data)[[2]]
    d <- list()
    # fit linear model and overwrite data with residuals
    sel <- S[, seq_len(k-1)]
    model <- lm(data ~ sel + 0, na.action=na.exclude)
    eps <- array(NA, dim(data))
    eps[!is.na(data)] <- model$residuals
    data <- ifelse(abs(eps) < 10^(-20), 0, eps)
    dimnames(data)[[2]] <- features
    data
}


R/simData.R
#' @title simData
#'
#' @description simulate Data with orthogonal feature clusters and replicated
#' samples. Each feature cluster corresponds to a different latent factor and
#' contains 10 redundant features. E.g. choosing samples = 100,
#' n_latent_factors = 5 and replicates = 2
#' will simulate a 50 x 200 data matrix, where the first 100 samples belong to
#' replicate 1 and sample 101-200 belong to replicate 2.
#' \cr
#' \cr
#'
#'
#' @details
#' simData constructs n_latent_factors by generating a random matrix
#' \eqn{\mathbf{Q}} whose row vectors
#' \eqn{\mathbf{Q}_{i\cdot}  \sim \mathcal{N}(0,1)}
#' with \eqn{n} samples and \eqn{i \in \{1, \dots, \textrm{n\_latent\_factors}\}} are
#' orthonormal, each corresponding to a different latent factor. To simulate a
#' set of redundant feature groups, it generates 10 features \eqn{X_{j\cdot}}
#' for each latent factor \eqn{\mathbf{Q}_{i\cdot}} by scaling each latent
#' factor by a random factor \eqn{\delta_j \sim \mathcal{N}(0,1)} and adding
#' replicate specific noise
#' \eqn{\pmb{\epsilon}_c \sim \mathcal{N}(0,0.1)} with
#' \eqn{c \in \{1, \dots, \textrm{replicates}\}} preserving orthogonality.
#'
#' @param samples number of samples to generate
#' @param n_latent_factors number of latent factors to generate
#' @param replicates number of replicates to generate
#'
#' @return \code{SummarizedExperiment} object carrying simulated data, with 
#' \code{colData} indicating which sample belongs to which replicate
#'
#' @importFrom MASS mvrnorm
#' @importFrom pracma gramSchmidt
#' @importFrom stats rnorm
#' @importFrom SummarizedExperiment SummarizedExperiment
#' @importFrom methods is
#' 
#' @export
#'
#' @examples
#' # simulate data with 100 samples, 20 features generated by 2 latent factors
#' # and 2 replicates
#' simData(samples=100, n_latent_factors=2, replicates=2)
simData <- function(samples, n_latent_factors, replicates) {
    stopifnot(
        is.numeric(n_latent_factors),
        is.numeric(samples),
        is.numeric(replicates),
        length(n_latent_factors) == 1,
        length(samples) == 1,
        length(replicates) == 1
    )
    
    # total number of features
    p <- n_latent_factors * 10
    # total number of measurements
    n <- samples * replicates
    # sample target feature clusters from
    # standard multivariate normal distribution
    m <- rep(10, n_latent_factors)
    # randomly sample targetcluster vectors
    target <- mvrnorm(samples, mu=m,
                            diag(rep(1, n_latent_factors)))
    # store magnitude of vector
    mag <- sqrt(colSums(target^2))
    # orthonormalize the vectors
    gs <- gramSchmidt(target)
    target <- gs$Q
    # rescale orthonormal vectors to original size
    target <- t(t(target) * mag)
    base <- array(dim=c(samples, p))
    # generate 9 additional linearly dependent features
    # for each feature cluster
    for (i in seq_len(n_latent_factors)){
        base[, (i-1)*10+1] <- target[, i]
        for (j in seq_len(9)){
            coefs <- rnorm(1)
            # add noise to reduce orthogonality between features in same cluster
            base[, (i-1)*10+1+j] <- (abs(coefs) * target[, i]) +
                rnorm(samples, 0, 0.1)
        }
    }
    # generate replicates and add noise
    replist <- list()
    for (i in seq_len(replicates)) {
        # add different noise to each replicate and concatenate
        noise <- array(rnorm(nrow(base) * ncol(base), mean=0,
                        sd=0.5), dim(base))
        if (i == 1){
            rep <- base
        } else {
            rep <- base + noise
        }
        replist[[i]] <- rep
    }
    # combine replicates
    data <- Reduce(function(x, y) rbind(x, y), replist)

    featnames <- c()
    for(i in seq_len(n_latent_factors)){
        featnames <- c(featnames, paste0("Latent_factor", i, "_Feature", seq_len(10)))
    }
    dimnames(data)[[2]] <-  featnames
    reps <- data.frame(replicates=rep(seq_len(replicates), each=samples))

    se <- SummarizedExperiment(assays=list(data=t(data)), colData=reps)
    se
}


R/FeatSeekR.R


#' FeatSeekR an R package for unsupervised feature selection
#' 
#' FeatSeekR performs unsupervised feature selection using replicated 
#' measurements. It iteratively selects features with the highest 
#' reproducibility across replicates, after projecting out those dimensions 
#' from the data that are spanned by the previously selected features. The 
#' selected a set of features has a high replicate reproducibility and a high 
#' degree of uniqueness.
#' 
#' For information on how to use this package please type 
#' \code{vignette("FeatSeekR-vignette")}.
#' 
#' Please post questions regarding the package to the Bioconductor Support Site:
#' 
#' \url{https://support.bioconductor.org}
#' 
#' @author Tümay Capraz
#' 
#' @docType package
#' @name FeatSeekR
#' @keywords package
NULL





#' @title FeatSeek
#' @description This function ranks features of a 2
#' dimensional array according to their reproducibility between replicates.
#'
#'
#' 
#' @param data \code{SummarizedExperiment} with assay named \code{data}, where each sample
#' belongs to a different replicate. Which sample belongs to which replicate
#' should indicated in \code{colData} slot replicates. Or \code{matrix} with features x samples.
#'
#' @param replicates numeric vector of length samples,
#' indicating which sample belongs to which replicate. Only required if \code{data} is
#' provided as \code{matrix}.
#' @param init \code{character vector} with names of initial features.
#' If \code{NULL} the feature with highest F-statistic will be used
#' @param max_features \code{integer} number of features to rank
#' @param verbose \code{logical} indicating whether messages should be printed
#' 
#' @return \code{SummarizedExperiment} containing one assay with the selected features.
#' \code{rowData} stores for each selected feature the F-statistic under \code{metric},
#' the cumulative explained variance under \code{explained_variance} and
#' the feature names under \code{selected}
#'
#' @examples
#' # run FeatSeek to select the top 20 features
#' data <-  array(rnorm(100*30), dim=c(30, 100),
#' dimnames <- list(paste("feature", seq_len(30)), NULL))
#' reps <- rep(c(1, 2), each=50)
#' res <- FeatSeek(data, reps, max_features=20)
#'
#' # res stores the 20 selected features ranked by their replicate reproducibility
#'
#' @importFrom SummarizedExperiment SummarizedExperiment assay
#' @importFrom methods is
#' 
#' @export
FeatSeek <- function(data, replicates=NULL, max_features=NULL, init=NULL, verbose=TRUE) {
    stopifnot(
        is.numeric(replicates) | is.null(replicates),
        is.numeric(max_features) | is.null(max_features),
        is.character(init) | is.null(init),
        is.logical(verbose)
    )
    se <- check_input(data, max_features, replicates)

    p <- nrow(se)
    n <- ncol(se)
    r <- length(unique(se$replicates))

    # initialize starting set of features
    init <- init_selected(init, se)
    if (verbose == TRUE){
        message("Input data has: \n",
                n, " samples \n",
                r, " replicates \n",
                p, " features")
    }


    # init max_features
    if(is.null(max_features)){
        max_features <- p
    }
    # initialize output dataframe
    res <-
        data.frame(
            metric = rep(NA, max_features),    # selection metric
            selected = rep("", max_features),  # name of selected features
            explained_variance = rep(NA, max_features)
            # variance explained by selected features
        )

    metric_all <- list()

    # initialize matrix of selected features
    S <- array(NA, dim=c(n, p))
    k <- 1

    data <- t(assay(se, "data"))
    replicates <- se$replicates
    data0 <- data
    # start ranking features
    # in each iteration we select the feature with
    # highest consistency between replicates
    # after projecting out the previously selected ones
    if (verbose == TRUE){
        message("Starting feature ranking!")
    }
    while (k <= max_features)  {
        if (k > 1) {
            data <- fit_lm(data, S, k)
        }
        # check if any replicate contains only zero residuals
        if (any(apply(data, 2, function(x) {all(x == 0)}))) {
            # if all residuals are close to zero stop selection procedure
            if (verbose == TRUE){
                message(
                    "Stopping selection procedure with k=",
                    k - 1,
                    " features because remaining features are linear combinations
                    of already selected features!"
                )
            }
            break
        }
        # calculate mean pairwise correlations between all replicates
        metric <- calcFstat(data, replicates)
        metric_all[[k]] <- metric
        names(metric) <- dimnames(data)[[2]]
        if (k > length(init)) {
            # select feature whose residuals have the highest correlation
            I <- names(metric)[which.min(metric)]
        } else{
            # first features are set to init ones
            I <- init[k]
        }
        # store metric of selected feature
        res$metric[k] <- metric[I]
        # store selected feature
        res$selected[k] <- I
        S[, k] <- data[, I, drop = FALSE]
        # drop selected feature from data
        data <- data[, dimnames(data)[[2]]!= I, drop = FALSE]
        if (verbose == TRUE){
            message("Iteration: ", k, " selected = ", res$selected[k],
                    ", replicate F-statistic = ", res$metric[k])
        }
        k <- k + 1

    }
    if (verbose == TRUE){
        message("Finished feature ranking procedure!")
        message("Calculating explained variance of selected feature sets!")
    }
    res <- res[seq_len(k-1), ]
    res$explained_variance <- vapply(seq_along(res$selected), function(i){
        variance_explained(data0, res$selected[seq_len(i)])
    }, numeric(1))
    se_res <- SummarizedExperiment(assays=list(selected=t(data0[, res$selected])), rowData=res)
    se_res
}



#' @title calcFstat
#'
#' @param data 2 dimensional \code{array} with samples x features, where each sample
#' belongs to a different replicate
#' @param reps \code{numeric vector} of length samples,
#' indicating which sample belongs to which replicate
#' @param scale \code{logical} whether to scale the data, default = TRUE
#'
#' @return F-statistic for all features
#'
#' @importFrom stats lm
#' @importFrom methods is
#' 
#' @keywords internal
calcFstat <- function(data, reps, scale=TRUE){
    stopifnot(
        is.numeric(data),
        is.numeric(reps),
        length(reps) == nrow(data),
        is.logical(scale)
    )
    data <- scale(data)
    f <- vapply(seq_len(dim(data)[2]), function(x){
        m <- lm(data[,x] ~ as.factor(reps))
        s <- summary(m)
        f <- s$fstatistic[1]
        }, numeric(1))
    f
}


R/check_input.R
#' @title check_input
#'
#' @description Checks input \code{data}. Input \code{data} should be a 2 dimensional \code{array}
#' with samples x features or \code{SummarizedExperiment} carrying one assay named \code{data}
#' and \code{colData} indicating which sample belongs to which replicate
#'
#'
#' @param data input \code{data} provided to \code{FeatSeek} either \code{SummarizedExperiment} or
#' 2 dimensional \code{array} with features x samples
#' @param replicates if \code{data} is a 2 dimensional \code{array} with features x samples
#' a vector indicating which sample corresponds to which replicate
#' must be provided
#'
#'
#' @return \code{SummarizedExperiment} where replicate information is stored in colData
#' 
#' @importFrom SummarizedExperiment SummarizedExperiment assay
#' @keywords internal
check_input <- function(data, max_features, replicates=NULL){
    if (!is(data,"SummarizedExperiment")){
        reps <- data.frame(replicates=replicates)
        if (!length(replicates) == ncol(data)) stop("Replicate indicator vector not same length as samples in data!")
        se <- SummarizedExperiment(assays=list(data=data), colData=reps)
    } else {
        se <- data
    }
    fnames <- rownames(se)
    if (all(vapply(fnames, function(x) is.null(x), logical(1)))) stop(
        "No feature names given or features not in correct dimension of data array!")

    if (!is.null(max_features) & max_features > nrow(se)) stop("Max features higher than features in data!")

    if( length(unique(se$replicates)) < 2) stop("At least 2 replicates required!")
    if (any(table(se$replicates) < 2)) stop("Not every sample has at least 2 replicates!")
    se
}

#' @title init_selected
#'
#' @description Checks if preselected init features are in input data.
#' If init is NULL, it is set to feature with highest replicate correlation.
#'
#' @param init preselected starting set of features
#' @param data input data as SummarizedExperiment
#'
#'
#'
#' @return names of initial set of feature
#'
#' @importFrom SummarizedExperiment assay
#' @importFrom stats lm
#' @importFrom methods is
#' 
#' @keywords internal
init_selected <- function(init, se){
    stopifnot(
        is(se, "SummarizedExperiment"),
        # 'n_features' should be a scalar integer and 
        # mustn't exeed the number of available features
        is.character(init) | is.null(init)
    )
    features <- dimnames(assay(se, "data"))[[1]]
    replicates <- se$replicates
    if (is.null(features)) stop("No feature names given or features not in correct dimension of data array!")
    # check if init features are in data
    if(!is.null(init) & !all(init %in% features)){
        stop("Could not find init features in data!")
    }

    # start with feature with highest replicate correlation if no init
    # features were given
    if (is.null(init)){
        data <- t(data.frame(assay(se, "data")))
        f <- vapply(seq_len(dim(data)[[2]]), function(x){
            m <- lm(replicates ~ data[, x])
            s <- withCallingHandlers(summary(m),
                                warning = function(w){
                                    if(startsWith(conditionMessage(w),
                                        "essentially perfect fit")){
                                        invokeRestart("muffleWarning")
                                    } else {
                                        message(w$message)
                                    }
                                })
            s$fstatistic[1]
        }, numeric(1))
        names(f) <-  features
        init <- names(which.min(f))
    }
    init
}





R/postprocessing.R
#' @title stop_criterion
#'
#' @description Get the optimal number of features based on the KS statistic
#' between the inital and the current distribution of pairwise replicate correlations.
#'
#' @param dist distribution of pairwise replicate pearson correlations
#'
#' @return returns optimal number of features
#'
#' @keywords internal
stop_criterion <- function(stat){
  for (i in seq(3,length(stat))){
    # get change in ks statistic as sliding window
    window <- ifelse(i <= 6, i-1, i-5)
    if (abs(stat[[i]]-stat[[window]]) < 0.005)
      break
  }
  stop <- ifelse(i <= 6, i-2, i-6)
  stop
}


#'
#'
#' #' @title Postprocessing
#' #' @description  After the selection procedure we determine the optimal minimum set of features.
#' #' The optimal set maximizes the fraction of variance of the remaining features
#' #' that can be explained by the selected set of features, while minimizing redundancy.
#' #' @param data 3 dimensional array with samples x replicates x features.
#' #' @param result result dataframe of feature selection run with FeatSeek
#' #' @param lambda regulates tradeoff between variance explained and redundancy
#' #'
#' #' @return list with original dataframe with variance explained and redundancy metrics for each selected feature added and optimal number of features
#' #' @export
#' #'
#' postprocessing <- function(data, result, lambda=0.5){
#'
#'     n <- dim(data)[1]
#'     r <- dim(data)[2]
#'     p <- dim(data)[3]
#'
#'     # check if data has feature names
#'     if(!is.null(dimnames(data)[[3]])) {
#'         features <- dimnames(data)[[3]]
#'     }else {
#'         stop("No feature names given or features not in correct dimension of data array!")
#'     }
#'
#'     if (!"selected" %in% names(result)){
#'         stop("No selected features in selection  result!\n Please pass output of FeatSeek as result argument.")
#'     }
#'     if (length(result$selected) < 2 ){
#'         stop("Number of selected features must be >= 2!")
#'     }
#'
#'     if (any(!result$selected %in% dimnames(data)[[3]])){
#'         stop("Selected features not in input data!")
#'     }
#'     # average over replicates
#'     mean_data <- apply(data, MARGIN=c(1,3), sum)/r
#'     # selected features
#'     selected_data <- mean_data[,result$selected]
#'
#'     result$var_explained <- rep(NA, dim(result)[1])
#'     result$redundancy <- rep(NA, dim(result)[1])
#'
#'     # First we calculate the fraction of variance explained by the selected feature set for each remaining feature.
#'     for (i in seq_len(length(result$selected))){
#'         sel <- selected_data[,result$selected[seq_len(i)]]
#'         rest <- mean_data[, !dimnames(mean_data)[[2]] %in% result$selected[1:i], drop=FALSE]
#'         # model remaining features by selected features
#'         model <- lm(rest ~  sel + 0)
#'         # model holds a fit for each remaining feature
#'         summary <- summary(model)
#'         # get average of adjusted r square values over all remaining features
#'         r_square <- mean(unlist(lapply(summary, function(x) {x$adj.r.squared})))
#'         result$var_explained[[i]] <- r_square
#'     }
#'
#'     # Next we check for redundancy in all possible feature sets of our feature ranking. T
#'     # We measure redundancy by SVD entropy. SVD entropy is defined as the normalized entropy of the singular values of a SVD of all selected features.
#'     # A feature set with low SVD entropy has higher redundancy as most of the variance can be captured by few singualar values.
#'
#'     entropy <- rep(NA, length(result$selected))
#'     for (i in seq_len(length(result$selected))){
#'         sel <- selected_data[,result$selected[seq_len(i)]]
#'         singv <- svd(sel)[[1]]
#'         singv.norm <- (singv^2) / (sum(singv^2))
#'         entropy[i] <- calc_entropy(singv.norm)
#'     }
#'     entropy[1] <- 1
#'     result$redundancy <- entropy
#'     # get optimum number of selected features which maximizes both variance explained of remaining features and SVD entropy
#'     # lambda controls the contribution of redundancy and variance explained
#'     opt <- which.max((result$var_explained*lambda) + result$redundancy*(1-lambda))
#'     # return optimal set
#'     list(result=result, opt_features = opt)
#' }


R/evaluate_features.R
#' @title variance_explained
#'
#' @param data 2 dimensional \code{array} samples x features 
#' @param selected \code{character vector} of selected features
#'
#' @return average variance explained by selected features
#'
#' @importFrom stats lm
#' @importFrom methods is
#' 
#' @keywords internal
variance_explained <- function(data, selected){
    stopifnot(
        is.numeric(data),
        is.character(selected),
        length(selected) <= ncol(data)
    )
    # calculate the fraction of variance explained by the selected feature set for each remaining feature.
    rest <- data
    data_sel <- data[, selected]
    # model remaining features by selected features
    model <- lm(rest ~ data_sel + 0)
    # model holds a fit for each remaining feature
    # catch warning that fit is perfect, as this is expected at certain number
    # of selected features
    s <- withCallingHandlers(summary(model),
                            warning = function(w){
                                if(startsWith(conditionMessage(w), "essentially perfect fit")){
                                    invokeRestart("muffleWarning")
                                } else {
                                    message(w$message)
                                }
                            })
    r <- mean(vapply(s, function(x){
        x$r.squared
    }, numeric(1)))

    r

}



R/plots.R
#' @title plotVarianceExplained
#' @description plot variance explained from 1 to max_features in res
#'
#' @param res result \code{SummarizedExperiment} from \code{FeatSeek} function
#'
#' @return returns plot of variance explained vs number of features
#'
#' @examples
#' # run FeatSeek to select the top 20 features
#' data <-  array(rnorm(100*30), dim=c(30,100),
#'             dimnames = list(paste("feature", seq_len(30)), NULL))
#' reps <- rep(c(1,2), each=50)
#' res <- FeatSeek(data, reps, max_features=20)
#'
#' # res stores the 20 selected features ranked by their replicate reproducibility
#' plotVarianceExplained(res)
#'
#' @importFrom SummarizedExperiment rowData
#' @importFrom methods is
#' @export
plotVarianceExplained <- function(res){
    stopifnot(
        is(res, "SummarizedExperiment")
    )
    if (is.null(rowData(res)$explained_variance)){
        stop("'res' does not contain attribute explained_variance. Please run FeatSeek first!")
    }

    plot(seq_len(nrow(res)), rowData(res)$explained_variance,
        xlab="Number of selected features",
        ylab="Fraction of explained variance")
}



#' @title plotSelectedFeatures
#'
#' @description plot correlation matrix of selected feature sets
#'
#' @param res result \code{SummarizedExperiment} from \code{FeatSeek} function
#' @param n_features top \code{n_features} to plot. if \code{NULL} then the maximum number
#'  of features in res will be plotted
#' @param assay assay slot to plot from result \code{SummarizedExperiment} object, 
#'  default is the selected features slot
#'
#' @return returns heatmap of selected features
#'
#' @examples
#' # run FeatSeek to select the top 20 features
#' data <-  array(rnorm(100*30), dim=c(30,100),
#'             dimnames = list(paste("feature", seq_len(30)), NULL))
#' reps <- rep(c(1,2), each=50)
#' res <- FeatSeek(data, reps, max_features=20)
#'
#' # res stores the 20 selected features ranked by their replicate reproducibility
#' # plot the top 5 features
#' plotSelectedFeatures(res, n_features=5)
#' 
#' @importFrom stats cor
#' @importFrom SummarizedExperiment assayNames assay
#' @importFrom pheatmap pheatmap
#' @importFrom methods is
#' 
#' @export
plotSelectedFeatures <- function(res, n_features=NULL, assay="selected"){
    stopifnot(
        is(res, "SummarizedExperiment"),
        # 'n_features' should be a scalar integer and 
        # mustn't exeed the number of available features
        is.numeric(n_features) | is.null(n_features), 
        # assay should be character that matches a unique assay name
        is.character(assay), length(assay) == 1,
        sum(grepl(assay, assayNames(res))) == 1
    )
    
    if (is.null(n_features)){
        n_features <- nrow(res)
    } else {
        stopifnot(
            length(n_features) == 1,
            round(n_features) == n_features
        )
    }
    cor <- cor(t(assay(res, assay))[, seq_len(n_features)], use = "pairwise.complete.obs")

    range <- max(abs(cor))
    pheatmap(cor, treeheight_row=0, treeheight_col = 0, legend=TRUE,
                    show_colnames=FALSE, show_rownames=TRUE,
                    breaks = seq(-range, range, length.out=100))

}
